File: xiang.li/1277.kt
class Solution {
    fun countSquares(matrix: Array<IntArray>): Int {
        if (matrix == null || matrix.isEmpty()) {
            return 0
        }
        val m = matrix.size
        val n: Int = matrix[0].size
        val dp = Array(m + 1) { IntArray(n + 1) }
        var res = 0
        for (i in 1..m) {
            for (j in 1..n) {
                if (matrix[i - 1][j - 1] == 1) {
                    dp[i][j] = 1 + Math.min(dp[i - 1][j], Math.min(dp[i - 1][j - 1], dp[i][j - 1]))
                }
                res += dp[i][j]
            }
        }
        return res
    }
}

File: hang.liu/1277.js
var countSquares = function(matrix) {
  var m = matrix.length;
  var n = matrix[0].length;
  var result = 0;

  for (var i = 0; i < m; i++) {
    for (var j = 0; j < n; j++) {
      if (matrix[i][j] > 0 && i > 0 && j > 0) {
        matrix[i][j] =
          Math.min(
            matrix[i - 1][j - 1],
            Math.min(matrix[i - 1][j], matrix[i][j - 1])
          ) + 1;
      }
      result += matrix[i][j];
    }
  }

  return result;
};
